connect/eaas/runner/managers/transformation.py

```diff
@@ -0,0 +1,275 @@
+            result_message.output.message = traceback.format_exc()[:4000]
+
+        return result_message
+
+    async def process_transformation(self, task_data, tfn_request, method):
+        input_file = await asyncio.get_running_loop().run_in_executor(
+            self.executor,
+            self.download_excel,
+            tfn_request,
+        )
+        output_file = NamedTemporaryFile(
+            suffix=tfn_request['files']['input']['name'].split('.')[-1],
+        )
+
+        read_queue = asyncio.Queue(TRANSFORMATION_TASK_MAX_PARALLEL_LINES)
+        write_queue = asyncio.Queue()
+
+        reader_task = asyncio.get_running_loop().run_in_executor(
+            self.executor,
+            self.read_excel,
+            input_file.name,
+            read_queue,
+        )
+        writer_task = asyncio.get_running_loop().run_in_executor(
+            self.executor,
+            self.write_excel,
+            output_file.name,
+            write_queue,
+            tfn_request['stats']['total'],
+            task_data.input.object_id,
+            task_data,
+        )
+        processor_task = asyncio.create_task(self.process_rows(
+            read_queue, write_queue, method, tfn_request['stats']['total'],
+        ))
+
+        await asyncio.gather([reader_task, writer_task, processor_task])
+        await self.send_output_file(task_data, output_file)
+        output_file.close()
+
+        return TransformationResponse.done()
+
+    def download_excel(self, tfn_request):
+        input_file_name = tfn_request['files']['input']['name']
+        input_file = NamedTemporaryFile(suffix=input_file_name.split('.')[-1])
+
+        with requests.get(
+            url=f'{self.config.get_api_url()}{input_file_name}',
+            stream=True,
+            headers={
+                'API_KEY': self.config.api_key,
+                **self.config.get_user_agent(),
+            },
+        ) as response:
+            response.raise_for_status()
+            for chunk in response.iter_content(chunk_size=DOWNLOAD_CHUNK_SIZE):
+                input_file.write(chunk)
+
+        return input_file
+
+    def read_excel(self, filename, queue):
+        wb = load_workbook(filename=filename, read_only=True)
+        ws = wb.worksheets[0]
+        for idx, row in enumerate(ws.rows, start=1):
+            asyncio.run_coroutine_threadsafe(
+                queue.put((idx, row)),
+                asyncio.get_running_loop(),
+            )
+
+        wb.close()
+
+    async def process_rows(self, read_queue, write_queue, method, total_rows):
+        rows_processed = 0
+        while rows_processed < total_rows:
+            row_idx, row = await read_queue.get()
+
+            if row_idx == 1:
+                await write_queue.put((row_idx, row))
+            elif inspect.iscoroutinefunction(method):
+                asyncio.create_task(self.async_process_row(
+                    method,
```

Comment:
I'm not 100% sure but maybe here we need to send the transformation settings as a parameter (also in the sync version). This settings are saved on the transformation->settings field, a field related to the transformation_request. This data is required in order to know what to do in the transformation function.

Reply:
the transformation request is available for the transformation method as `self.transformation_request` as you can see here: https://github.com/cloudblue/connect-eaas-core/blob/master/connect/eaas/core/extension.py#L330